### PR TITLE
SSIMULACRA 2

### PR DESCRIPTION
**Pull request by jonsneyers**
Major new version of the SSIMULACRA metric.
Changes compared to the original version:

- works in XYB color space
- uses 1-norm and 4-norm (instead of 1-norm and max-norm-after-downscaling)
- penalizes both smoothing and ringing artifacts (instead of only penalizing ringing but not smoothing)
- removed specific blockiness detection

The score weights were obtained by running Nelder-Mead simplex search, optimizing to minimize MSE and maximize Kendall and Pearson correlation for training data consisting of 17611 subjective quality scores, validated on separate validation data consisting of 4292 scores.

Results for the full dataset (almost 22k subjective scores):

Metric | Kendall correlation | Spearman | Pearson |
-- | -- | -- |-- 
PSNR | 0.34677 | 0.49926 | 0.48178
SSIM | 0.41612 | 0.58939 | 0.52543
SSIMULACRA | -0.5214 | -0.7124 | -0.6891
VMAF | 0.61720 | 0.81549 | 0.77781
Butteraugli 2-norm | -0.6572 | -0.8461 | -0.8103
SSIMULACRA 2 | 0.74185 | 0.91695 | 0.90244

Results for just the validation set (4292 subjective scores):

Metric | Kendall correlation | Spearman | Pearson |
-- | -- | -- |-- 
SNR | 0.35089 | 0.50092 | 0.50157
SSIM | 0.45002 | 0.62777 | 0.55710
SSIMULACRA | -0.5881 | -0.7845 | -0.7808
VMAF | 0.58799 | 0.78798 | 0.75031
Butteraugli 2-norm | -0.6226 | -0.8127 | -0.7834
SSIMULACRA 2 | 0.72163 | 0.90354 | 0.89504

Computing the mean absolute error between opinion scores (on a scale of 0 to 100) and SSIMULACRA 2 results, on the full set the MAE is 4.274 and on the validation set the MAE is 4.470.

### CODE REVIEW

File: tools/ssimulacra2.cc [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+  }
+}
+
+// Temporary storage for Gaussian blur, reused for multiple images.
+class Blur {
+ public:
+  Blur(const size_t xsize, const size_t ysize)
+      : rg_(jxl::CreateRecursiveGaussian(1.5)), temp_(xsize, ysize) {}
+
+  void operator()(const ImageF& in, ImageF* JXL_RESTRICT out) {
```
Comment (veluca93) [inline tools/ssimulacra2.cc:92]: Let's please not do overloaded `operator()` :) (arguably also not operator() at all, but that might be more OK)

File: tools/ssimulacra2.cc [hunk 2]
```diff
@@ -0,0 +1,435 @@
...
+      const float* JXL_RESTRICT row_s11 = s11.PlaneRow(c, y);
+      const float* JXL_RESTRICT row_s22 = s22.PlaneRow(c, y);
+      const float* JXL_RESTRICT row_s12 = s12.PlaneRow(c, y);
+      for (size_t x = 0; x < m1.xsize(); ++x) {
+        float mu1 = row_m1[x];
+        float mu2 = row_m2[x];
+        float mu11 = mu1 * mu1;
+        float mu22 = mu2 * mu2;
+        float mu12 = mu1 * mu2;
+        float nom_m = 2 * mu12 + kC1;
```
Comment (veluca93) [inline tools/ssimulacra2.cc:132]: nit: s/nom/num/

File: tools/ssimulacra2.cc [hunk 3]
```diff
@@ -0,0 +1,435 @@
...
+      const float* JXL_RESTRICT rowm2 = mu2.PlaneRow(c, y);
+      for (size_t x = 0; x < img1.xsize(); ++x) {
+        double d1 = (1.0 + std::abs(row2[x] - rowm2[x])) /
+                        (1.0 + std::abs(row1[x] - rowm1[x])) -
+                    1.0;
+        // d1 > 0: distorted has an edge where original is smooth
+        //         (indicating ringing, color banding, blockiness, etc)
+        // d1 < 0: original has an edge where distorted is smooth
+        //         (indicating smoothing, blurring, smearing, etc)
+        double d2 = std::max(d1, 0.0);
```
Comment (veluca93) [inline tools/ssimulacra2.cc:167]: call this smoothness, and line 172 blockiness?

File: tools/ssimulacra2.cc [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+  if (ssim > 0) {
+    ssim = 100.0 - 10.0 * pow(ssim, 0.6940375837127916);
+
+  } else {
+    ssim = 100.0;
+  }
+  return ssim;
+}
+
+void BminusY(jxl::Image3F& img) {
```
Comment (veluca93) [inline tools/ssimulacra2.cc:345]: Misleading name :)

File: tools/ssimulacra2.cc [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+
+The weights were obtained by running Nelder-Mead simplex search,
+optimizing to minimize MSE and maximize Kendall and Pearson correlation
+for training data consisting of 17611 subjective quality scores,
+validated on separate validation data consisting of 4292 scores.
+*/
+double Msssim::Score() const {
+  double ssim = 0.0;
+
+  const double weight[108] = {-0.4887721343447775,
```
Comment (veluca93) [inline tools/ssimulacra2.cc:206]: static constexpr

File: tools/ssimulacra2.cc [hunk 6]
```diff
@@ -0,0 +1,435 @@
...
+
+  Msssim msssim = ComputeSSIMULACRA2(orig, dist);
+  printf("%.8f\n", msssim.Score());
+  return 0;
+}
+
+}  // namespace
+}  // namespace ssimulacra2
+
+int main(int argc, char** argv) { return ssimulacra2::Run(argc, argv); }
```
Comment (veluca93) [inline tools/ssimulacra2.cc:435]: main and run really should be in a different file if the intention is to allow using this as a library.

If not, then the .h file is not particularly useful :)

Comment (jonsneyers) [reply tools/ssimulacra2.cc:435]: Right. I guess having it as a library could be useful (say to be used in benchmark_xl), so I'll split it up.

File: tools/ssimulacra2.cc (hunk 3 above)
Comment (jonsneyers) [reply tools/ssimulacra2.cc:167]: Something descriptive could in any case be useful, yes.

What about 'detail_lost' and 'artifacts' or something like that?

Comment (veluca93) [reply tools/ssimulacra2.cc:167]: Sure

File: tools/ssimulacra2.cc (hunk 1 above)
Comment (jonsneyers) [reply tools/ssimulacra2.cc:92]: I don't disagree, but this part of the code was just copied from ssimulacra.cc, so it probably should be changed in both places (or be put in some shared auxiliary .h or something).

**Comment by jonsneyers** [comment]
Added a separate main file, added some more info to the help screen, and made it work for images with alpha too.

File: tools/ssimulacra2.cc [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+   compressed with JPEG, JPEG 2000, JPEG XL, WebP, AVIF, and HEIC.
+*/
+
+#include "tools/ssimulacra2.h"
+
+#include <stdio.h>
+
+#include <cmath>
+
+//#include "lib/extras/codec.h"
```
Comment (veluca93) [inline tools/ssimulacra2.cc:32]: remove?

**Review by veluca93 — APPROVED** [review]

**Review by veluca93 — APPROVED** [review]

**Comment by jonsneyers** [comment]
> @jonsneyers I see that you are using coefficients correlations to evaluate metrics. You might be interested to learn that there is a new coefficient correlation in town, with very interesting properties and therefore maybe that this new meta-metric could help you compare and improve compressions metrics? https://news.ycombinator.com/item?id=29687613

Thanks for the pointer! Sounds interesting indeed.

Note however that most metrics require a nonlinear mapping to be used as MOS predictors, so in general, rank correlation (like Kendall's tau) is more useful than Pearson correlation to compare the performance of various metrics. But I have to read that paper...

**Comment by LifeIsStrange** [comment]
I'm not an expert but I found this visualization quite impressive https://twitter.com/adad8m/status/1474754752193830912?s=21
